Error building old-time (due to hsc)

This is the hsc2hs invocation done:
```sh
external/my_ghc/bin/hsc2hs external/haskell_old_time/System/Time.hsc -o /tmp/x.hs -Iexternal/haskell_old_time/include -Ibazel-out/k8-fastbuild/genfiles/external/haskell_old_time/include
```

old-time doesn't include HsTime.h unless it's on GHC: https://github.com/haskell/old-time/blob/master/System/Time.hsc#L110. I can manually run the command with `-D__GLASGOW_HASKELL__` and it works.

```python
load("@io_tweag_rules_haskell//haskell:haskell.bzl", "haskell_library")

genrule(
    name = "config",
    srcs = [
        "configure",
        "install-sh",
        "config.sub",
        "config.guess",
        "include/HsTime.h",
        "include/HsTimeConfig.h.in",
    ],
    outs = ["include/HsTimeConfig.h"],
    cmd = """
        cd `dirname $(location configure)`
        ./configure
        DIR=`pwd`
        cd -
        cp $$DIR/include/HsTimeConfig.h $@
    """,
)

cc_library(
    name = "cbits",
    srcs = [
        "include/HsTime.h",
        ":config",
    ],
    includes = ["include"],
)

haskell_library(
    name = "old-time",
    srcs = glob(["System/Time.hsc"]),
    prebuilt_dependencies = ["base"],
    visibility = ["//visibility:public"],
    deps = [":cbits"],
)
```

Yes, that fixes it. Now I'm stuck on MissingH, but that's not related to this issue.

```
external/haskell_MissingH/src/System/Time/Utils.hs:35:1: error:
    System.Time: Can't be safely imported!
    The module itself isn't safe.
```

--- a/haskell/BUILD
+++ b/haskell/BUILD
@@ -2,6 +2,7 @@ package(default_visibility = ["//visibility:public"])
 
 exports_files([
   "cc.bzl",
+  "ghc-defs-cleanup.sh",
   "ghc-version-check.sh",
   "haddock.bzl",
   "haskell.bzl",

--- a/haskell/ghc-defs-cleanup.sh
+++ b/haskell/ghc-defs-cleanup.sh
@@ -1,0 +1,13 @@
+#!/usr/bin/env bash
+#
+# Usage: ghc-defs-cleanup.SH <ORIGINAL-FILE> <OUTPUT-FILE>
+
+# This is what Cabal sets:
+#
+# https://github.com/haskell/cabal/blob/master/Cabal/Distribution/Simple/PreProcess.hs#L530-L537
+#
+# so we should be OK too.
+
+grep "#define __GLASGOW_HASKELL__" $1 >  $2
+grep "#define .*_HOST_OS"          $1 >> $2
+grep "#define .*_BUILD_ARCH"       $1 >> $2

--- a/haskell/haskell.bzl
+++ b/haskell/haskell.bzl
@@ -57,6 +57,10 @@ _haskell_common_attrs = {
     default = "1.0.0",
     doc = "Library/binary version. Internal - do not use."
   ),
+  "_ghc_defs_cleanup": attr.label(
+    allow_single_file = True,
+    default = Label("@io_tweag_rules_haskell//haskell:ghc-defs-cleanup.sh"),
+  )
 }
 
 def _haskell_binary_impl(ctx):

--- a/haskell/hsc2hs.bzl
+++ b/haskell/hsc2hs.bzl
@@ -7,6 +7,8 @@ load(":path_utils.bzl",
 
 load(":tools.bzl", "get_hsc2hs")
 load(":cc.bzl", "cc_headers")
+load("@bazel_skylib//:lib.bzl", "paths")
+load(":tools.bzl", "get_compiler")
 
 def hsc_to_hs(ctx):
   """Process all hsc files into Haskell source files.
@@ -17,19 +19,63 @@ def hsc_to_hs(ctx):
   Returns:
     list of File: New Haskell source files to use.
   """
+  ghc_defs_dump = _make_ghc_defs_dump(ctx)
   sources = []
   for f in ctx.files.srcs:
     if f.extension == "hsc":
-      sources.append(_process_hsc_file(ctx, f))
+      sources.append(_process_hsc_file(ctx, ghc_defs_dump, f))
     else:
       sources.append(f)
   return sources
 
-def _process_hsc_file(ctx, hsc_file):
+def _make_ghc_defs_dump(ctx):
+  """Generate a file containing GHC default pre-processor definitions.
+
+  Args:
+    ctx: Rule context.
+
+  Returns:
+    File: The file with GHC definitions.
+  """
+  raw_filename      = "ghc-defs-dump-{0}-{1}.hs".format(ctx.attr.name, ctx.attr.version)
+  dummy_src         = ctx.actions.declare_file(raw_filename)
+  ghc_defs_dump_raw = ctx.actions.declare_file(paths.replace_extension(raw_filename, ".hspp"))
+  ghc_defs_dump     = ctx.actions.declare_file(paths.replace_extension(raw_filename, ".h"))
+
+  ctx.actions.write(dummy_src, "")
+  args = ctx.actions.args()
+  args.add([
+    "-E",
+    "-optP-dM",
+    "-cpp",
+    dummy_src.path,
+  ])
+
+  ctx.actions.run(
+    inputs     = [dummy_src],
+    outputs    = [ghc_defs_dump_raw],
+    executable = get_compiler(ctx),
+    arguments  = [args],
+  )
+
+  ctx.actions.run(
+    inputs     = [ghc_defs_dump_raw],
+    outputs    = [ghc_defs_dump],
+    executable = ctx.file._ghc_defs_cleanup,
+    arguments  = [
+      ghc_defs_dump_raw.path,
+      ghc_defs_dump.path,
+    ],
+  )
+
+  return ghc_defs_dump
+
+def _process_hsc_file(ctx, ghc_defs_dump, hsc_file):
   """Process a single hsc file.
 
   Args:
     ctx: Rule context.
+    ghc_defs_dump: File with GHC definitions.
     hsc_file: hsc file to process.
 
   Returns:
@@ -48,9 +94,12 @@ def _process_hsc_file(ctx, hsc_file):
   # Bring in scope the header files of dependencies, if any.
   hdrs, include_args = cc_headers(ctx)
   args.add(include_args)
+  args.add("-I{0}".format(ghc_defs_dump.dirname))
+  args.add("-i{0}".format(ghc_defs_dump.basename))
 
   ctx.actions.run(
-    inputs = depset(transitive = [depset(hdrs), depset([hsc_file])]),
+    inputs = depset(transitive =
+                    [depset(hdrs), depset([hsc_file, ghc_defs_dump])]),
     outputs = [hs_out, hsc_output_dir],
     use_default_shell_env = True,
     progress_message = "hsc2hs {0}".format(hsc_file.basename),

--- a/tests/hsc/Foo.hsc
+++ b/tests/hsc/Foo.hsc
@@ -1,6 +1,8 @@
 module Foo (hscFiredFoo) where
 
+#if __GLASGOW_HASKELL__ >= 700
 #ifndef _INTERNAL_HSC_DO_NOT_DEFINE_ME
 hscFiredFoo :: String
 hscFiredFoo = "hscFiredFoo"
+#endif
 #endif

